Tweak test configuration to run in extra_tests/

diff --git a/extra_tests/hpy_tests/_vendored/conftest.py b/extra_tests/hpy_tests/_vendored/conftest.py
--- a/extra_tests/hpy_tests/_vendored/conftest.py
+++ b/extra_tests/hpy_tests/_vendored/conftest.py
@@ -2,17 +2,12 @@ import pytest
 from .support import ExtensionCompiler
 from hpy.debug.pytest import hpy_debug # make it available to all tests
 
-def pytest_addoption(parser):
-    parser.addoption(
-        "--compiler-v", action="store_true",
-        help="Print to stdout the commands used to invoke the compiler")
-
 @pytest.fixture(scope='session')
 def hpy_devel(request):
     from hpy.devel import HPyDevel
     return HPyDevel()
 
-@pytest.fixture(params=['cpython', 'universal', 'debug'])
+@pytest.fixture(params=['universal', 'debug'])
 def hpy_abi(request):
     return request.param
 
diff --git a/extra_tests/hpy_tests/_vendored/debug/__init__.py b/extra_tests/hpy_tests/_vendored/debug/__init__.py
new file mode 100644
